server/api/signup.js

```diff
@@ -1,51 +1,83 @@
-const Express = require('express');
-const Crypto = require('crypto');
+const {checkUsernameRequirements, checkPasswordRequirements} = require("../../utilities/AccountValidators");
 
-var db = require('../db');
+const Crypto = require('crypto');
+const Sqlite3 = require('sqlite3');
+var db = require('../db').db;
 
-/* POST /signup
- *
- * This route creates a new user account.
- *
- * A desired `username` and `password` are submitted to this route via an HTML form. 
- * The password is hashed and then a new user record is inserted into the 
- * database.  If the record is successfully created, the user is logged in.
+/**
+ * The callback function for the /api/signup POST route. Signs up a user into the database.
  * 
- * This was also shamelessly stolen from the passport tutorial.
+ * The `res` parameter should contain both `username` and `password` in its body.
+ * @param {Express.Request} req - The Request object generated by the route.
+ * @param {Express.Response} res - The Result object generated by the route.
+ * @param {Function} next - The next() function used in the callback.
+ * @returns Nothing.
  */
 function signup(req, res, next) {
-    if (!req.body.password || !req.body.username) {
-      res.status(400).send("Error: Username or password not set!");
-      return next("Error: Username or password not set!");
+  // use test database if test flag is set
+  if (req.body.test) {
+    db = new Sqlite3.Database("signup_test_db2.db")
+  }
+
+  // check that signup data is real
+  if (!req.body.username || !req.body.password) {
+    res.status(400).send("Error: Username or password not set!");
+    return next("Error: Username or password not set!");
+  }
+
+  // test username and password; should throw an error if there's an issue
+  try {
+    checkUsernameRequirements(req.body.username);
+    checkPasswordRequirements(req.body.password);
+  } catch (error) {
+    res.status(401).send(error.message);
+    return next(error.message);
```

Comment:
i think returning `next()` makes it so the endpoint doesn't actually respond to the request. I'd just return `res.status...`